Zend/zend_vm_def.h

```diff
@@ -4548,11 +4548,13 @@ ZEND_VM_HANDLER(161, ZEND_GENERATOR_RETURN, CONST|TMP|VAR|CV, ANY, SPEC(OBSERVER
 
 	ZEND_OBSERVER_FCALL_END(generator->execute_data, &generator->retval);
 
-	EG(current_execute_data) = EX(prev_execute_data);
+	zend_execute_data *prev_execute_data = EX(prev_execute_data);
 
 	/* Close the generator to free up resources */
 	zend_generator_close(generator, 1);
 
+	EG(current_execute_data) = prev_execute_data;
```

Comment:
Won't this changes the behavior and make generators behave differently from regular functions?
See ``zend_leave_helper``. It sets ``EG(current_execute_data)`` before ``zend_clean_and_cache_symbol_table`` and other cleanup calls...

Reply:
You are right. However, I don't know what the correct behavior is then.

```
function bar() {
    yield from foo();
}

function foo() {
    $foo = new Foo();
    yield;
}
```

In this case, resuming generator `bar()` does not truly resume generator `bar()`, it resumes generator `foo()`. When popping the stack frame, we're left with the `execute_fake` frame that is inserted for `bar()` as a placeholder, which has `func` set to `NULL`, which causes this crash. Instead of just popping one stack frame, we could pop two. However, that leaves us with the output:

```
#0 %s(%d): Foo->__destruct()
```

If you think that is more correct, then I'm fine with it.

@bwoebi Could you comment on what you think the right output would be?

Reply:
The _consistent_ output would be:
```
#0 %s(%d): Foo->__destruct()
#1 %s(%d): bar()
```

But I do agree that we should change stack traces everywhere to include the parent frame, but too late for 8.4 I guess.

Reply:
This is targeted to PHP-8.2!

Reply:
I think we may need a similar change in `zend_dispatch_try_catch_finally_helper`:

``` php
<?php

class Foo {
    public function __destruct() {
        debug_print_backtrace();
    }
}

function bar() {
    yield from foo();
}

function foo() {
    $foo = new Foo();
    yield throw new Exception();
}

$gen = bar();
foreach ($gen as $dummy);
```

An alternative might be to change `zend_fetch_debug_backtrace()` so that we enter into `zend_generator_check_placeholder_frame()`.

Reply:
@dstogov Better?

Reply:
yes. I think this is better.